design/pipeline-caching.md

```diff
@@ -0,0 +1,157 @@
+
+Other more elaborate caching mechanisms may also be available such as virtualized file systems with prefetch which may provide significant performance boosts particularly in sparse file access scenarios but introduce other issues which would need to make that approach strictly opt-in.
+
+The caching strategy would default to  dedup file transfer mechanism but allow an option to override the caching strategy with another approach, for example"
+
+Selecting the cache strategy would simply be a matter of specifying a strategy on the task inputs (for both the restore and save tasks, restore shown below):
+
+```yaml
+steps:
+- task: RestoreCache@0
+  inputs:
+    strategy: dedup
+    key: |
+      package.json
+      yarn.lock
+    paths: |
+      node_modules
+```
+
+### YAML syntax
+
+The example task usage above uses the explicit task references. We also want to provide a streamlined syntax for caching, applied to the example above.
+
+```yaml
+steps:
+- restoreCache: yarn.lock
+- script: yarn
+- saveCache:
+  key: yarn.lock
+  paths: node_modules
+- script: yarn build
+```
+
+This more compact syntax just reduces some of the noise around using the caching task. Additionally we are planning to integrate caching into the the ```use:``` syntax. This will look like the following:
+
+```yaml
+steps:
+- use: node
+  cache: true
+- script: |
+    yarn
+    yarn build
+    yarn test
+```
+
+The idea behind the ```use:``` syntax is that given what we are using (in this case ```node```) we make some automatic decisions about what and how to cache. This logic will use various heuristics to make the decision about what to cache, for example, it will look for the presence of a ```yarn.lock``` file or a ```package-lock.json``` to determine where the ```node_modules``` path may be located.
+
+Other ```use:``` statements will have similar heuristics to pick the best defaults.
+
+Note that the ```use:``` syntax will inject the cache save step at the end of the build process which will not always be desirable. In those cases developers can fall back to specifying the ```restorCache:``` and ```saveCache:``` YAML statements.
+
+### Cache Scoping
+
+Getting scoping right is important for maximising cache hits and but also avoiding the cache becoming an attack vector to insert malicous code into official/master builds.
+
+As a result we will automatically scope caches to the branch that they are running against. The caches will also be hierarchical so a feature branch will be able to get a hit on the master, but when populatingn the cache in the ```saveCache:``` step, the contents of that cache won't be used on the master build.
+
+For PR builds, the PR build will use the cache of the branch it is merging into or from with preference been given to the branch it is merging from. At this stage, PR builds, regardless of whether they are from a branch within the repo or from a forked clone will not be able to store content in the cache.
```

Comment:
When a PR includes a change to cached dependencies, (i.e., update package version, add package, etc), what will the mechanics of this be?  i.e., is this treated as a cache miss, or perhaps the cache is restored, but still requires a conditional step to `yarn` or `npm i`, etc.

Reply:
It will be treated as a cache miss.